File: deployment-services/shared/resilience/circuit_breaker.py
import time
import enum
import logging
from typing import Callable, Any, Optional

logger = logging.getLogger("resilience")

class CircuitState(enum.Enum):
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"

class CircuitBreakerOpenError(Exception):
    def __init__(self, name: str):
        super().__init__(f"Circuit breaker [{name}] is OPEN – call rejected")
        self.name = name

class CircuitBreaker:
    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        success_threshold: int = 2,
        timeout: float = 30.0,
        half_open_max_calls: int = 3,
        fallback: Optional[Callable[[], Any]] = None
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout
        self.half_open_max_calls = half_open_max_calls
        self.fallback = fallback
        
        self.state = CircuitState.CLOSED
        self.failures = 0
        self.successes = 0
        self.total_calls = 0
        self.last_failure_time = None
        self.last_state_change = time.time()
        self.half_open_calls = 0

    def execute(self, func: Callable[..., Any], *args, **kwargs) -> Any:
        self.total_calls += 1
        self._maybe_transition_from_open()

        if self.state == CircuitState.OPEN:
            logger.warning(f"Circuit breaker [{self.name}] is OPEN – rejecting call")
            if self.fallback:
                return self.fallback()
            raise CircuitBreakerOpenError(self.name)

        if self.state == CircuitState.HALF_OPEN:
            if self.half_open_calls >= self.half_open_max_calls:
                logger.warning(f"Circuit breaker [{self.name}] HALF_OPEN – max probe calls reached, rejecting")
                if self.fallback:
                    return self.fallback()
                raise CircuitBreakerOpenError(self.name)
            self.half_open_calls += 1

        try:
            result = func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure(e)
            raise

    def get_state(self) -> CircuitState:
        self._maybe_transition_from_open()
        return self.state

    def reset(self):
        self._transition_to(CircuitState.CLOSED)
        self.failures = 0
        self.successes = 0
        self.half_open_calls = 0

    def _on_success(self):
        self.failures = 0
        if self.state == CircuitState.HALF_OPEN:
            self.successes += 1
            if self.successes >= self.success_threshold:
                self._transition_to(CircuitState.CLOSED)

    def _on_failure(self, err: Exception):
        self.last_failure_time = time.time()
        self.successes = 0
        self.failures += 1

        if self.state == CircuitState.HALF_OPEN:
            self._transition_to(CircuitState.OPEN)
            return

        if self.state == CircuitState.CLOSED and self.failures >= self.failure_threshold:
            self._transition_to(CircuitState.OPEN)

        logger.error(f"Circuit breaker [{self.name}] recorded failure: {err}")

    def _maybe_transition_from_open(self):
        if (
            self.state == CircuitState.OPEN and
            self.last_failure_time is not None and
            time.time() - self.last_failure_time >= self.timeout
        ):
            self._transition_to(CircuitState.HALF_OPEN)

    def _transition_to(self, next_state: CircuitState):
        if self.state == next_state:
            return
        logger.info(f"Circuit breaker [{self.name}] transition: {self.state.value} -> {next_state.value}")
        self.state = next_state
        self.last_state_change = time.time()
        self.half_open_calls = 0
        self.successes = 0
        if next_state == CircuitState.CLOSED:
            self.failures = 0

File: identity-services/packages/common/src/resilience/circuit-breaker.ts
import { CreateLogger } from "../utils/logger.js";

const rootLogger = CreateLogger({ name: "resilience" });

export type CircuitState = "CLOSED" | "OPEN" | "HALF_OPEN";

export interface CircuitBreakerOptions {
    /** Number of consecutive failures before opening the circuit */
    failureThreshold?: number;
    /** Number of consecutive successes in HALF_OPEN before closing */
    successThreshold?: number;
    /** Milliseconds to wait in OPEN state before transitioning to HALF_OPEN */
    timeout?: number;
    /** Max concurrent calls allowed in HALF_OPEN state */
    halfOpenMaxCalls?: number;
    /** Optional fallback function called when circuit is OPEN */
    fallback?: <T>() => T | Promise<T>;
}

export interface CircuitMetrics {
    state: CircuitState;
    failures: number;
    successes: number;
    totalCalls: number;
    lastFailureTime: number | null;
    lastStateChange: number;
}

export class CircuitBreakerOpenError extends Error {
    constructor(name: string) {
        super(`Circuit breaker [${name}] is OPEN – call rejected`);
        this.name = "CircuitBreakerOpenError";
    }
}

export class CircuitBreaker {
    private state: CircuitState = "CLOSED";
    private failures = 0;
    private successes = 0;
    private totalCalls = 0;
    private lastFailureTime: number | null = null;
    private lastStateChange = Date.now();
    private halfOpenCalls = 0;

    private readonly failureThreshold: number;
    private readonly successThreshold: number;
    private readonly timeout: number;
    private readonly halfOpenMaxCalls: number;
    private readonly fallback?: <T>() => T | Promise<T>;
    private readonly log: ReturnType<typeof rootLogger.child>;

    constructor(
        private readonly name: string,
        options: CircuitBreakerOptions = {}
    ) {
        this.failureThreshold = options.failureThreshold ?? 5;
        this.successThreshold = options.successThreshold ?? 2;
        this.timeout = options.timeout ?? 30_000;
        this.halfOpenMaxCalls = options.halfOpenMaxCalls ?? 3;
        this.fallback = options.fallback;
        this.log = rootLogger.child({ circuitBreaker: name });
    }

    async execute<T>(fn: () => Promise<T>): Promise<T> {
        this.totalCalls++;
        this.maybeTransitionFromOpen();

        if (this.state === "OPEN") {
            this.log.warn({ state: this.state }, "Circuit OPEN – rejecting call");
            if (this.fallback) return this.fallback<T>();
            throw new CircuitBreakerOpenError(this.name);
        }

        if (this.state === "HALF_OPEN") {
            if (this.halfOpenCalls >= this.halfOpenMaxCalls) {
                this.log.warn("Circuit HALF_OPEN – max probe calls reached, rejecting");
                if (this.fallback) return this.fallback<T>();
                throw new CircuitBreakerOpenError(this.name);
            }
            this.halfOpenCalls++;
        }

        try {
            const result = await fn();
            this.onSuccess();
            return result;
        } catch (err) {
            this.onFailure(err);
            throw err;
        }
    }

    getState(): CircuitState {
        this.maybeTransitionFromOpen();
        return this.state;
    }

    getMetrics(): CircuitMetrics {
        return {
            state: this.getState(),
            failures: this.failures,
            successes: this.successes,
            totalCalls: this.totalCalls,
            lastFailureTime: this.lastFailureTime,
            lastStateChange: this.lastStateChange,
        };
    }

    reset() {
        this.transitionTo("CLOSED");
        this.failures = 0;
        this.successes = 0;
        this.halfOpenCalls = 0;
    }

    private onSuccess() {
        this.failures = 0;
        if (this.state === "HALF_OPEN") {
            this.successes++;
            if (this.successes >= this.successThreshold) {
                this.transitionTo("CLOSED");
            }
        }
    }

    private onFailure(err: unknown) {
        this.lastFailureTime = Date.now();
        this.successes = 0;
        this.failures++;

        if (this.state === "HALF_OPEN") {
            this.transitionTo("OPEN");
            return;
        }

        if (this.state === "CLOSED" && this.failures >= this.failureThreshold) {
            this.transitionTo("OPEN");
        }

        this.log.error({ err, failures: this.failures }, "Circuit breaker recorded failure");
    }

    private maybeTransitionFromOpen() {
        if (
            this.state === "OPEN" &&
            this.lastFailureTime !== null &&
            Date.now() - this.lastFailureTime >= this.timeout
        ) {
            this.transitionTo("HALF_OPEN");
        }
    }

    private transitionTo(next: CircuitState) {
        if (this.state === next) return;
        this.log.info({ from: this.state, to: next }, "Circuit breaker state transition");
        this.state = next;
        this.lastStateChange = Date.now();
        this.halfOpenCalls = 0;
        this.successes = 0;
        if (next === "CLOSED") this.failures = 0;
    }
}
